rust-demangle.c

```diff
@@ -1147,27 +1152,170 @@ static void demangle_const_str_literal(struct rust_demangler *rdm) {
+
+                uint32_t c = 0;
+                for (size_t i = 0; i < hex.nibbles_len; i++)
+                    c = (c << 4) | decode_hex_nibble(hex.nibbles[i]);
+
+                if (!(c < 0xd800 || (c > 0xdfff && c < 0x10ffff))) {
+                    break; // Not a valid unicode scalar
+                }
+
+                if (c >= 0x20 && c <= 0x7e) {
+                    // Printable ASCII
+                    char v = (char)c;
+                    print_str(rdm, &v, 1);
+                } else {
+                    // FIXME show printable unicode characters without hex
+                    // encoding
+                    PRINT("\\u{");
+                    char s[9] = {0};
+                    sprintf(s, "%" PRIx32, c);
+                    PRINT(s);
+                    PRINT("}");
+                }
+            }
+
+            ident.ascii += escape_len + 2;
+            ident.ascii_len -= escape_len + 2;
+        } else {
+            bool found = false;
+            for (size_t i = 0; i < ident.ascii_len; i++) {
+                if (ident.ascii[i] == '$' || ident.ascii[i] == '.') {
+                    print_str(rdm, ident.ascii, i);
+                    ident.ascii += i;
+                    ident.ascii_len -= i;
+                    found = true;
+                    break;
+                }
+            }
+            if (!found) {
+                break;
+            }
+        }
+    }
+
+    print_str(rdm, ident.ascii, ident.ascii_len);
+}
+
+static void demangle_legacy_path(struct rust_demangler *rdm) {
+    bool first = true;
+
+    while (1) {
+        if (eat(rdm, 'E')) {
+            // FIXME Maybe check if at end of symbol?
+            return;
+        }
+
+        struct rust_mangled_ident name = parse_ident(rdm);
+
+        if (!rdm->verbose && peek(rdm) == 'E' && is_rust_hash(name)) {
+            // Skip printing the hash if verbose mode is disabled.
+            eat(rdm, 'E');
+            break;
+        }
+
+        if (!first) {
+            PRINT("::");
+        }
+        first = false;
+
+        print_legacy_ident(rdm, name);
+
+        CHECK_OR(!rdm->errored, return);
+    }
+}
+
 bool rust_demangle_with_callback(
     const char *mangled, int flags,
     void (*callback)(const char *data, size_t len, void *opaque), void *opaque
 ) {
-    // Rust symbols always start with R, _R or __R.
-    if (mangled[0] == '_' && mangled[1] == 'R')
-        mangled += 2;
-    else if (mangled[0] == 'R')
-        // On Windows, dbghelp strips leading underscores, so we accept "R..."
-        // form too.
-        mangled += 1;
-    else if (mangled[0] == '_' && mangled[1] == '_' && mangled[2] == 'R')
-        // On OSX, symbols are prefixed with an extra _
-        mangled += 3;
-    else
-        return false;
-
-    // Paths always start with uppercase characters.
-    if (!IS_UPPER(mangled[0]))
-        return false;
-
     struct rust_demangler rdm;
 
     rdm.sym = mangled;
```

Comment:
Since now the structure of `rust_demangle_with_callback` is such that it initializes a `rust_demangler` and then operates on it, I realized that it there could be a separate `demangle_symbol` entry-point that takes a freshly built `rust_demangler` and does everything else (doesn't have to be in this PR tho, if it's even a good idea).

Really, my only concern could be accidentally using `mangled` instead of `rdm.sym` but that could also be fixed by renaming `mangled` to `whole_mangled_symbol` or something idk.

Reply:
I've renamed it.